blockchain/types/tx_internal_data_dynamic_fee.go

```diff
@@ -0,0 +1,512 @@
+	d := newTxInternalDataDynamicFee()
+
+	d.AccountNonce = nonce
+	d.Recipient = to
+	d.GasLimit = gasLimit
+
+	if len(data) > 0 {
+		d.Payload = common.CopyBytes(data)
+	}
+
+	if amount != nil {
+		d.Amount.Set(amount)
+	}
+
+	if gasTipCap != nil {
+		d.GasTipCap.Set(gasTipCap)
+	}
+
+	if gasFeeCap != nil {
+		d.GasFeeCap.Set(gasFeeCap)
+	}
+
+	if accessList != nil {
+		copy(d.AccessList, accessList)
+	}
+
+	if chainID != nil {
+		d.ChainID.Set(chainID)
+	}
+
+	return d
+}
+
+func newTxInternalDataDynamicFeeWithMap(values map[TxValueKeyType]interface{}) (*TxInternalDataDynamicFee, error) {
+	d := newTxInternalDataDynamicFee()
+
+	if v, ok := values[TxValueKeyNonce].(uint64); ok {
+		d.AccountNonce = v
+		delete(values, TxValueKeyNonce)
+	} else {
+		return nil, errValueKeyNonceMustUint64
+	}
+
+	if v, ok := values[TxValueKeyTo].(common.Address); ok {
+		d.Recipient = &v
+		delete(values, TxValueKeyTo)
+	} else {
+		return nil, errValueKeyToMustAddress
+	}
+
+	if v, ok := values[TxValueKeyAmount].(*big.Int); ok {
+		d.Amount.Set(v)
+		delete(values, TxValueKeyAmount)
+	} else {
+		return nil, errValueKeyAmountMustBigInt
+	}
+
+	if v, ok := values[TxValueKeyData].([]byte); ok {
+		d.Payload = common.CopyBytes(v)
+		delete(values, TxValueKeyData)
+	} else {
+		return nil, errValueKeyDataMustByteSlice
+	}
+
+	if v, ok := values[TxValueKeyGasLimit].(uint64); ok {
+		d.GasLimit = v
+		delete(values, TxValueKeyGasLimit)
+	} else {
+		return nil, errValueKeyGasLimitMustUint64
+	}
+
+	if v, ok := values[TxValueKeyGasFeeCap].(*big.Int); ok {
+		d.GasFeeCap.Set(v)
+		delete(values, TxValueKeyGasFeeCap)
+	} else {
+		return nil, errValueKeyGasFeeCapMustBigInt
+	}
+	if v, ok := values[TxValueKeyGasTipCap].(*big.Int); ok {
+		d.GasTipCap.Set(v)
+		delete(values, TxValueKeyGasTipCap)
+	} else {
+		return nil, errValueKeyGasTipCapMustBigInt
+	}
+	if v, ok := values[TxValueKeyAccessList].(AccessList); ok {
+		d.AccessList = make(AccessList, len(v))
+		copy(d.AccessList, v)
+		delete(values, TxValueKeyAccessList)
+	} else {
+		return nil, errValueKeyAccessListInvalid
+	}
+
+	if v, ok := values[TxValueKeyChainID].(*big.Int); ok {
+		d.ChainID.Set(v)
+		delete(values, TxValueKeyChainID)
+	} else {
+		return nil, errValueKeyChainIDInvalid
+	}
+
+	if len(values) != 0 {
+		for k := range values {
+			logger.Warn("unnecessary key", k.String())
+		}
+		return nil, errUndefinedKeyRemains
+	}
+
+	return d, nil
+}
+
+func (t *TxInternalDataDynamicFee) Type() TxType {
+	return TxTypeDynamicFee
+}
+
+func (t *TxInternalDataDynamicFee) GetRoleTypeForValidation() accountkey.RoleType {
+	return accountkey.RoleTransaction
+}
+
+func (t *TxInternalDataDynamicFee) GetAccountNonce() uint64 {
+	return t.AccountNonce
+}
+
+func (t *TxInternalDataDynamicFee) GetPrice() *big.Int {
+	return t.GasFeeCap
```

Comment:
Returning GasFeeCap is fine?
How about implementing GetEffectiveGasPrice in here?

Reply:
In go-eth, This function return gasFeeCap. I think this is because `GasFeeCap` is required when calculating the cost to execute a transaction in TxPool.

Reply:
Um.... i could not understand why `GasFeeCap` is required when calculating the cost in Txpool?

Reply:
Txpool checks whether the tx sender is able to pay the tx fee and the sending amount. Because `GasFeeCap` is the maximum affordable gas price, txpool checks the payability with the value in a conservative way.

Reply:
okay i undrestand. Then I understand that the `GasPrice` function of `TxInternalDataDynamicFee` is used only for TxPool to check affordable to pay with maximum amount.

Reply:
Yes. When the tx fee is calculated during the tx execution, `effectivateGasUsed` should be used normally.